x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/source_fields_merging/strategies/merge_all_fields_with_source.test.ts

```diff
@@ -1523,14 +1525,113 @@ describe('merge_all_fields_with_source', () => {
       const fields: SignalSourceHit['fields'] = {
         foo: [{ bar: ['single_value'], zed: ['single_value'] }],
       };
-      const doc: SignalSourceHit = { ...emptyEsResult(), _source, fields };
+      const doc: SignalSourceHit = { ...emptyEsResult(), _source: cloneDeep(_source), fields };
       const merged = mergeAllFieldsWithSource({ doc, ignoreFields: [] })._source;
       expect(merged).toEqual<ReturnTypeMergeFieldsWithSource>({
         foo: [{ bar: ['single_value'], zed: ['single_value'] }],
       });
     });
   });
 
+  describe('mixed flattened and nested fields for the _source', () => {
+    test('merges fields into source if it is empty', () => {
+      const _source: SignalSourceHit['_source'] = {
+        'email.headers': {},
+      };
+
+      const fields: SignalSourceHit['fields'] = {
+        'email.headers.x-test': ['from fields'],
+      };
+      const doc: SignalSourceHit = { ...emptyEsResult(), _source: cloneDeep(_source), fields };
+      const merged = mergeAllFieldsWithSource({ doc, ignoreFields: [] })._source;
+      expect(merged).toEqual<ReturnTypeMergeFieldsWithSource>({
+        email: {
+          headers: {
+            'x-test': 'from fields',
+          },
+        },
+        // there is leftover of the empty flattened object left out, since we can't create wildcard
+        // map of string based paths to array paths in buildFieldsKeyAsArrayMap
+        'email.headers': {},
+      });
+    });
+
+    test('merges fields into source if it is "undefined"', () => {
```

Comment:
Similarly here, favor being more declarative with these descriptions; the more descriptive, the less likely that someone misunderstands/changes/invalidates the test later on: 
```suggestion
    test('drops a conflicting key if the value is undefined', () => {
```